src/main/java/ortus/boxlang/runtime/services/HttpService.java

```diff
@@ -0,0 +1,458 @@
+/**
+ * [BoxLang]
+ *
+ * Copyright [2023] [Ortus Solutions, Corp]
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package ortus.boxlang.runtime.services;
+
+import java.net.Authenticator;
+import java.net.InetSocketAddress;
+import java.net.PasswordAuthentication;
+import java.net.ProxySelector;
+import java.net.http.HttpClient;
+import java.security.KeyStore;
+import java.security.SecureRandom;
+import java.time.Duration;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.TimeUnit;
+
+import javax.net.ssl.KeyManagerFactory;
+import javax.net.ssl.SSLContext;
+
+import ortus.boxlang.runtime.BoxRuntime;
+import ortus.boxlang.runtime.async.executors.BoxExecutor;
+import ortus.boxlang.runtime.logging.BoxLangLogger;
+import ortus.boxlang.runtime.net.BoxHttpClient;
+import ortus.boxlang.runtime.scopes.Key;
+import ortus.boxlang.runtime.types.Array;
+import ortus.boxlang.runtime.types.IStruct;
+import ortus.boxlang.runtime.types.Struct;
+import ortus.boxlang.runtime.types.exceptions.BoxRuntimeException;
+import ortus.boxlang.runtime.util.EncryptionUtil;
+
+/**
+ * This service manages all HTTP clients in BoxLang.
+ * It is responsible for creating, caching, and reusing HTTP clients based on their configuration.
+ * It also handles the lifecycle events of the HTTP service within the BoxLang runtime.
+ */
+public class HttpService extends BaseService {
+
+	/**
+	 * Concurrent map that stores all HTTP reusable clients
+	 */
+	private final ConcurrentMap<Key, BoxHttpClient>	clients						= new ConcurrentHashMap<>();
```

Comment:
I'm not sure I'm really on board with the caching of clients here.  What happens if two threads request the same client and one calls a method on the client.  That thread is modifying by reference ( and modifying a runtime singleton ).   I see lots of potential problems with the cache implementation and see it having potentially time consuming maintenance iterations.  

If the developer want's to create a client with the same specifications for similar requests, then that's what functions are for.

Reply:
Clients are thread safe as per the HTTP spec.  You build a connection to a server. If you don't, you keep creating tcp/ip connections to the endpoints and not reusing the http pipes that are created.

The requests are also thread safe because they are transient.  So I don't see how this can cause contention, since the HTTP Client in the JDK actually suggests you treat them as singletons.

There are two concerns here
1. The http client connection
2. Requests to the http client

The http client connection SHOULD always be a singleton and it's thread safe inheriently.
The requests are transients and are built that way.

So in summary, I don't really know what concurrency issues you are referring to.

Reply:
From their Java Docs

java.net.http.HttpClient is thread-safe and designed to be reused across many requests and many threads.

* HttpClient is immutable after you build it (via HttpClient.newHttpClient() or HttpClient.newBuilder()...build()).
* It is safe to share a single instance across threads and across your whole app.

That’s the intended usage pattern: create one (or a small fixed number) and reuse it, instead of creating a new client per request.

The client is created according to connection details which have nothing to do with endpoint URLs.  Since we alwas get a new HttpRequest.newBuilder(). Then all our calls are thread safe.

Reply:
OK.  I think boundary between the internal class and the outer class confused me here.  Looking at `BoxHTTPClient` without the internal class, I can see how that works fine as a runtime singleton.  Disregard.